File: src/main/pairing/pairing-manager.ts
import { randomBytes } from 'node:crypto';
import type { PairingStore } from './pairing-store';

export const TOKEN_BYTE_LENGTH = 32;

export class PairingManager {
  constructor(private readonly store: PairingStore) {}

  async getDesktopId(): Promise<string> {
    return (await this.store.load()).desktopId;
  }
}

export function createToken(byteLength: number = TOKEN_BYTE_LENGTH): string {
  return randomBytes(byteLength).toString('base64url');
}

File: src-dotnet/SwitchifyPc.Core/Pairing/PairingManager.cs
using System.Security.Cryptography;

namespace SwitchifyPc.Core.Pairing;

public sealed class PairingManager
{
    private readonly IPairingStore store;

    public PairingManager(IPairingStore store)
    {
        this.store = store;
    }

    public async Task<string> GetDesktopIdAsync(CancellationToken cancellationToken = default)
    {
        return (await store.LoadAsync(cancellationToken)).DesktopId;
    }
}

public static class PairingToken
{
    public const int TokenByteLength = 32;

    public static string CreateToken(int byteLength = TokenByteLength)
    {
        return Base64UrlEncode(RandomNumberGenerator.GetBytes(byteLength));
    }

    private static string Base64UrlEncode(byte[] value)
    {
        return Convert.ToBase64String(value)
            .TrimEnd('=')
            .Replace("+", "-", StringComparison.Ordinal)
            .Replace("/", "_", StringComparison.Ordinal);
    }
}
